### DISCUSSION TITLE
finding nearest tag of a commit

### DISCUSSION THREAD

**Discussion by mainrs** [libgit2/libgit2#5589] [General] [answered]
Hey!

I need to find the closest tag for a given commit:
```text
commit a (HEAD -> master)
commit b
commit c (tag: v1.2.0)
```

Given the commit hash `b` I need to somehow be able to tell that the latest tag from that point of view was `v1.2.0`. Could you point me into directions on how to do this? 

Thank you!

**Comment by mainrs** [comment, marked as answer]
I figured it out. Maybe it's just me but I find the documentation of `libgit2` super hard to read. With no prior knowledge of how exactly `git` works under the hood and what mechanisms can be used to achieve certain things, it takes a relatively long time to figure stuff out :)

I could luckily modify my problem domain. I now had to just get a list of all commits between `HEAD` and the latest git tag. This can be done fairly easily without any tricks. It's worth noting though that the latest git tag does not always mean the latest version. People often enough re-tag old releases and those tags are newer but might point to an older version of the software. So I parsed the version strings manually using `semver` to get the highest, e.g. latest version tag. Getting the commit for a tag can be done by simply peeling it.

__Edit:__ For further readers, I've used the Rust library [git2](https://lib.rs/crates/git2) and not `libgit2` directly. So code won't help you probably.

**Reply by ltt-stripe** [reply]
Would be great if you could share some example code!

**Reply by mainrs** [reply]
> Never be the guy who finds a solution for his own issue and just says "Nevermind, found the solution, bye".

I did edit my comment and said how I've done it. I just retrieved a list of all commits between `HEAD` and the latest git tag. And to get the latest git tag, I've retrieved the list of all git tags and sorted them after parsing them by semver, not by date. Since, in a lot of cases, people re-tag older releases. 

The code won't help you, because I didn't use `libgit2`. I've used the Rust bindings. And since this is a C library, Rust code won't help you (probably).
